### ISSUE TITLE
Support microsecond resolution

### ISSUE THREAD

**Issue by msqr** [fboulnois/pg_uuidv7#12]
Thank you for writing this handy extension! I wondered if you would consider adding a function to create a v7 UUID variant with microsecond precision, by making use of the high 10 bits of bytes 6 & 7 for a 0-999 microseconds value instead of random bits. The reason for this is because I found the millisecond-level precision in a v7 UUID inadequate for an application I worked on, and using microseconds was a good compromise between time precision and randomness.

The UUID layout looks like this:

```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                           unix_ts_ms                          |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|          unix_ts_ms           |  ver  |       micros      |rnd|
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|var|                        rand_b                             |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                            rand_b                             |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```

I thought this could be an additional function like `uuid_generate_v7_micro()` so it did not impact the existing `uuid_generate_v7()` function. I have been using [a SQL function](https://github.com/SolarNetwork/solarnetwork-central/blob/2.42.2/solarnet-db-setup/postgres/postgres-init-common.sql#L299-L324) for this, but a C function would be much faster!

**Comment by fboulnois** [comment]
As I outline [in my comments on the RFC](https://github.com/ietf-wg-uuidrev/rfc4122bis/issues/101), I don't really like the alternate pseudorandom constructions. I'd also like to keep this extension as simple as possible so that there is only one way to build v7 UUIDs. As a result it is unlikely I will implement this.

**Comment by msqr** [comment]
Understood, thanks. I also hadn't noticed that the v7 draft has advanced since I started working with it, so for me I'm glad to see an optional additional time precision provision has been added to the proposal.

**Comment by msqr** [comment]
I appreciate you said you were unlikely to add the optional extra time precision support into this extension, but I thought I'd have a go at trying add this, and I thought the change could be quite focused with the addition of a default/optional argument to the `generate_uuid_v7()` function like `generate_uuid_v7(extra_ts_precision int = 0)`. The extra time precision is genuinely useful in many use cases, and would allow users to opt-in as desired.

I started playing around and was able to [get something started along these lines](https://github.com/SolarNetwork/pg_uuidv7/commit/c0c7bdb89756830625abb9e671945bb72ab088e3#diff-3fe3ac8f1e9abed54262e4400201e6de7f934e0bfb9ea4d6be9d816cb678d2f8), just for testing purposes, but my Postgres extension experience is limited and thought you might have more thoughts on the matter.

**Comment by fboulnois** [comment]
> I started playing around and was able to [get something started along these lines](https://github.com/SolarNetwork/pg_uuidv7/commit/c0c7bdb89756830625abb9e671945bb72ab088e3#diff-3fe3ac8f1e9abed54262e4400201e6de7f934e0bfb9ea4d6be9d816cb678d2f8), just for testing purposes, but my Postgres extension experience is limited and thought you might have more thoughts on the matter.

It might be easier to initially extract microseconds and then split it into millis and micros, but otherwise your approach seems reasonable to me.

**Comment by msqr** [comment]
I've flushed out the fractional millisecond support now, including updating the [README](https://github.com/SolarNetwork/pg_uuidv7#pg_uuidv7-use-the-new-v7-uuids-in-postgres). I also added some basic tests (run via `make installcheck`). I'm not sure the code is as efficient as it could be, or bug free, but it does seem to be performing well in my testing.

If you ever change your mind and would be interested in a PR, I'm happy to create one.
